File: java/webflux-security-example/src/main/java/com/van1164/webfluxsecurityexample/security/ReactiveOAuth2UserServiceImpl.java
package com.van1164.webfluxsecurityexample.security;

import com.van1164.webfluxsecurityexample.user_info.PrincipalDetails;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.oauth2.client.userinfo.DefaultReactiveOAuth2UserService;
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
import org.springframework.security.oauth2.client.userinfo.ReactiveOAuth2UserService;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.core.user.OAuth2User;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
@Slf4j
public class ReactiveOAuth2UserServiceImpl implements ReactiveOAuth2UserService<OAuth2UserRequest, OAuth2User> {
    // 구글로 부터 받은 userRequest 데이터에 대한 후처리되는 함수
    // 함수 종료시 @AuthenticationPrincipal 어노테이션이 만들어진다.
    @Override
    public Mono<OAuth2User> loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
        DefaultReactiveOAuth2UserService delegate = new DefaultReactiveOAuth2UserService();

        return delegate.loadUser(userRequest)
                .doOnNext(oAuth2User -> {
                    log.info("Enter OAuth2UserService");
                })
                .flatMap(oAuth2User -> {
                    if (userRequest.getClientRegistration().getRegistrationId().equals("google")) {
                        String email = oAuth2User.getAttributes().get("email").toString();
                        User user = new User(email, "", java.util.Collections.emptyList());
                        return Mono.just(new PrincipalDetails(user, oAuth2User.getAttributes()));
                    } else {
                        log.info("Unsupported OAuth");
                        return Mono.empty();
                    }
                });
    }
}

File: kotlin/webflux-security-example/src/main/kotlin/com/van1164/webfluxsecurityexample/security/ReactiveOAuth2UserServiceImpl.kt
package com.van1164.webfluxsecurityexample.security

import com.van1164.webfluxsecurityexample.logger
import com.van1164.webfluxsecurityexample.user_info.PrincipalDetailsReactive
import org.springframework.security.core.userdetails.User
import org.springframework.security.oauth2.client.userinfo.DefaultReactiveOAuth2UserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.client.userinfo.ReactiveOAuth2UserService
import org.springframework.security.oauth2.core.OAuth2AuthenticationException
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono


@Service
class ReactiveOAuth2UserServiceImpl : ReactiveOAuth2UserService<OAuth2UserRequest,OAuth2User> {
    //구글로 부터 받은 userRequest 데이터에 대한 후처리되는 함수
    //함수 종료시 @AuthenticationPrincipal 어노테이션이 만들어진다.
    @Throws(OAuth2AuthenticationException::class)
    override fun loadUser(userRequest: OAuth2UserRequest): Mono<OAuth2User> {
        val delegate = DefaultReactiveOAuth2UserService()

        return delegate.loadUser(userRequest)
            .doOnNext {
                logger.info { "Enter OAuth2UserService" }
            }
            .flatMap { oAuth2User ->
                when (userRequest.clientRegistration.registrationId) {
                    "google" -> {
                        val attributes = oAuth2User.attributes
                        val email = attributes["email"].toString()
                        val user = User(
                            email,
                            "",
                            listOf()
                        )
                        Mono.just(PrincipalDetailsReactive(user,attributes))
                    }
                    else -> {
                        logger.info{"Unsupported OAuth"}
                        Mono.empty<PrincipalDetailsReactive>()
                    }
                }
            }

    }


}